addons/sale_expense/models/hr_expense_sheet.py

```diff
@@ -15,42 +18,66 @@ def _compute_sale_order_count(self):
-                sol.order_id, sol.product_id, sol.product_uom_qty, sol.price_unit, sol.name, sol.id
-            FROM sale_order_line AS sol
-            WHERE sol.is_expense = TRUE
-                AND sol.order_id IN %s
-                AND sol.product_id IN %s
-                AND sol.product_uom_qty IN %s
-                AND sol.price_unit IN %s
-                AND sol.name IN %s
-            ORDER BY sol.order_id, sol.product_id, sol.product_uom_qty, sol.price_unit, sol.name
+              WITH aml(key_id, key_count, order_id, product_id, product_uom_qty, price_unit, name) AS (VALUES %s)
+            SELECT ARRAY_AGG(sol.id ORDER BY sol.id), aml.key_count
+              FROM aml,
+                   sale_order_line AS sol
+              JOIN res_company AS company ON sol.company_id = company.id
+              JOIN res_currency AS company_currency ON company.currency_id = company_currency.id
+         LEFT JOIN res_currency AS currency ON sol.currency_id = currency.id
+             WHERE sol.is_expense = TRUE
+               AND sol.order_id = aml.order_id
+               AND sol.product_id = aml.product_id
+               AND sol.product_uom_qty = aml.product_uom_qty
+               AND sol.name = aml.name
+               AND ROUND(sol.price_unit::numeric, COALESCE(currency.rounding, company_currency.rounding)::int)
+                   = ROUND(aml.price_unit::numeric, COALESCE(currency.rounding, company_currency.rounding)::int)
+               GROUP BY aml.key_id, aml.key_count
         """
-        self.env.cr.execute(query, (sale_order_ids, product_ids, quantities, tuple(set(aml_sol_unit_price_map.values())), names))
-        potential_sols_map = {
-            (row['order_id'], row['product_id'], row['product_uom_qty'], row['price_unit'], row['name']): row['id']
-            for row in self.env.cr.dictfetchall()
-        }
-        expensed_amls_keys = set(expensed_amls.mapped(
-            lambda aml: (aml.expense_id.sale_order_id.id, aml.product_id.id, aml.quantity, aml_sol_unit_price_map[aml.id], aml.name)
-        ))
-        return self.env['sale.order.line'].browse(sol_id for key, sol_id in potential_sols_map.items() if key in expensed_amls_keys)
+
+        # Get the keys used to fetch the corresponding sale order lines, and the number of times they are used
+        # We need the occurrences count to filter out the sale order lines so that we keep exactly one per expense
+        expense_keys_counter = Counter(expensed_amls.mapped(lambda aml: (
+            aml.expense_id.sale_order_id.id,
+            aml.product_id.id,
+            aml.quantity,
+            aml.currency_id.round(aml._sale_get_invoice_price(aml_to_so_map[aml.id])),
+            aml.name,
+        )))
+        expensed_amls_keys_and_count = tuple(
+            (key_id, key_count, *key, key_count) for key_id, (key, key_count) in enumerate(expense_keys_counter.items())
```

Comment:
How are the tests passing?
```suggestion
            (key_id, key_count, *key) for key_id, (key, key_count) in enumerate(expense_keys_counter.items())
```

Reply:
No idea, seems like Postgres is more flexible than it should :shrug: Fixed and repushed (and then repushed after a rebase)

Reply:
> Note that an AS clause is required when VALUES is used in a FROM clause, just as is true for SELECT. It is not required that the AS clause specify names for all the columns, but it's good practice to do so. (The default column names for VALUES are column1, column2, etc. in PostgreSQL, but these names might be different in other database systems.)

Today I learned
https://www.postgresql.org/docs/current/sql-values.html